types/src/transaction/authenticator.rs

```diff
@@ -737,16 +802,353 @@ impl fmt::Display for AuthenticationKey {
     }
 }
 
+#[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
+pub struct MultiKeyAuthenticator {
+    public_keys: MultiKey,
+    signatures: Vec<AnySignature>,
+    signatures_bitmap: aptos_bitvec::BitVec,
+}
+
+impl MultiKeyAuthenticator {
+    pub fn new(public_keys: MultiKey, signatures: Vec<(u8, AnySignature)>) -> Result<Self> {
+        ensure!(
+            public_keys.len() < (u8::MAX as usize),
+            "Too many public keys in MultiKeyAuthenticator."
+        );
+        let mut signatures_bitmap = aptos_bitvec::BitVec::with_num_bits(public_keys.len() as u16);
+        let signatures = signatures
+            .into_iter()
+            .map(|(idx, signature)| {
+                signatures_bitmap.set(idx as u16);
+                signature
+            })
+            .collect();
+
+        Ok(MultiKeyAuthenticator {
+            public_keys,
+            signatures,
+            signatures_bitmap,
+        })
+    }
+
+    pub fn verify<T: Serialize + CryptoHash>(&self, message: &T) -> Result<()> {
+        ensure!(
+            self.signatures_bitmap.count_ones() as usize == self.signatures.len(),
+            "Mismatch in signatures and signatures_bitmap"
+        );
+        println!(
+            "{} {}",
+            self.signatures.len(),
+            self.public_keys.signatures_required()
+        );
+        ensure!(
+            self.signatures.len() >= self.public_keys.signatures_required() as usize,
+            "Not enough signatures for verification"
+        );
+        for (idx, signature) in
+            std::iter::zip(self.signatures_bitmap.iter_ones(), self.signatures.iter())
+        {
+            signature.verify(&self.public_keys.public_keys[idx], message)?;
+        }
+        Ok(())
+    }
+
+    pub fn public_key_bytes(&self) -> Vec<u8> {
+        self.public_keys.to_bytes()
+    }
+
+    pub fn signature_bytes(&self) -> Vec<u8> {
+        bcs::to_bytes(&(&self.signatures, &self.signatures_bitmap))
+            .expect("Only unhandleable errors happen here.")
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct MultiKey {
+    public_keys: Vec<AnyKey>,
+    signatures_required: u8,
+}
+
+impl MultiKey {
+    pub fn new(public_keys: Vec<AnyKey>, signatures_required: u8) -> Self {
+        Self {
+            public_keys,
+            signatures_required,
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.public_keys.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.public_keys.len()
+    }
+
+    pub fn signatures_required(&self) -> u8 {
+        self.signatures_required
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
+        bcs::to_bytes(&self).expect("Only unhandleable errors happen here.")
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct SingleKeyAuthenticator {
+    public_key: AnyKey,
```

Comment:
Maybe call it `AnyPublicKey` or maybe even just `PublicKey`?

Reply:
will name AnyPublicKey